crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+    pub stream_hash: RwLock<HashMap<u32, pw::stream::Stream>>,
+    pub stream_listener: RwLock<HashMap<u32, pw::stream::StreamListener<i32>>>,
 }
 
 impl PwBackend {
-    pub fn new(streams: Arc<RwLock<Vec<Stream>>>) -> Self {
-        Self { streams }
+    pub fn new(stream_params: Arc<RwLock<Vec<Stream>>>) -> Self {
+        pw::init();
+
+        let thread_loop = Arc::new(PwThreadLoop::new(Some("Pipewire thread loop")).unwrap());
+        let get_loop = thread_loop.get_loop();
+
+        thread_loop.lock();
+
+        let context = pw::Context::new(&get_loop).expect("failed to create context");
+        thread_loop.start();
+        let core = context.connect(None).expect("Failed to connect to core");
+
+        // Create new reference for the variable so that it can be moved into the closure.
+        let thread_clone = thread_loop.clone();
+
+        // Trigger the sync event. The server's answer won't be processed until we start the thread loop,
+        // so we can safely do this before setting up a callback. This lets us avoid using a Cell.
+        let pending = core.sync(0).expect("sync failed");
+        let _listener_core = core
+            .add_listener_local()
+            .done(move |id, seq| {
+                if id == PW_ID_CORE && seq == pending {
+                    thread_clone.signal();
+                }
+            })
+            .register();
+
+        thread_loop.wait();
+        thread_loop.unlock();
+
+        println!("pipewire backend running");
+
+        Self {
+            stream_params,
+            thread_loop,
+            core,
+            context,
+            stream_hash: RwLock::new(HashMap::with_capacity(NR_STREAMS)),
+            stream_listener: RwLock::new(HashMap::with_capacity(NR_STREAMS)),
+        }
     }
 }
 
 impl AudioBackend for PwBackend {
-    fn write(&self, stream_id: u32) -> Result<()> {
-        log::trace!("PipewireBackend write stream_id {}", stream_id);
-        _ = std::mem::take(&mut self.streams.write().unwrap()[stream_id as usize].buffers);
+    fn write(&self, _stream_id: u32) -> Result<()> {
+        Ok(())
+    }
+
+    fn read(&self, _stream_id: u32) -> Result<()> {
+        log::trace!("PipewireBackend read stream_id {}", _stream_id);
         Ok(())
     }
 
-    fn read(&self, _id: u32) -> Result<()> {
-        log::trace!("PipewireBackend read stream_id {}", _id);
+    fn set_parameters(&self, stream_id: u32, mut msg: ControlMessage) -> Result<()> {
+        let descriptors: Vec<Descriptor> = msg.desc_chain.clone().collect();
+        let desc_request = &descriptors[0];
+        let request = msg
+            .desc_chain
+            .memory()
+            .read_obj::<VirtioSndPcmSetParams>(desc_request.addr())
+            .unwrap();
+        {
+            let stream_clone = self.stream_params.clone();
+            let mut stream_params = stream_clone.write().unwrap();
+            let st = &mut stream_params[stream_id as usize];
```

Comment:
use get_mut instead

Reply:
is there any significant difference between using &mut and get_mut()?

Reply:
Actually, my poor brain thought that only `std::ops::Index` existed and not `std::ops::IndexMut`. but the main difference is using `[index]` would panic if the index is out of bounds, where `get_mut` would return `Option<&mut T>` which might be useful here to at least log a proper "failure error" with `get_mut(stream_id as usize).expect("Stream must exist to set it parameters")`

Reply:
ok make sense, thanks